### DISCUSSION TITLE
NER support for multiple labels on the same span of text.

### DISCUSSION THREAD

**Discussion by narayanacharya6** [explosion/spaCy#8223] [Help: Coding & Implementations] [answered]
I have multiple NER models trained on different datasets which are then assembled into a single spaCy model. What I have noticed is that if a particular span was identified by the first NER model in the assembled pipeline, the span is no longer labeled by the subsequent NER models in the assembled pipeline. To better explain the situation please consider the following:

1. Train `ner_1` and `ner_2` separately.
2. Assemble both trained NERs into one pipeline as `["ner_1", "ner_2"]`.
3. Assemble another pipeline as `["ner_2", "ner_1"]`.
4. Run the 4 NER models, `ner_1`, `ner_2`, `["ner_1", "ner_2"]` and `["ner_2", "ner_1"]`, on a sentence, say, `Hydrogen is the lightest element on Earth and Iron Man is the coolest Avenger.`.

Example outputs:
1. For `ner_1`: `[('Hydrogen', 'LABEL_1'), ('Iron', 'LABEL_1')]`
2. For `ner_2`: `[('Iron Man', 'LABEL_2'), ('Avenger', 'LABEL_2')]`
3. For `["ner_1", "ner_2"]`: `[('Hydrogen', 'LABEL_1'), ('Iron', 'LABEL_1'), ('Avenger', 'LABEL_2')]`
4. For `["ner_2", "ner_1"]`: `[('Hydrogen', 'LABEL_1'), ('Iron Man', 'LABEL_2'), ('Avenger', 'LABEL_2')]`

I would expect the output for the assembled pipeline to be the union of the individual NER pipes, but that does not seem to be the case. It seems the first NER in the pipeline takes precedence and once a span is labeled, it is no longer allowed to be labeled by following pipes. Anything I can do to work around this issue for assembled pipelines?

Side note: I have read the discussion [here](https://github.com/explosion/spaCy/discussions/3961) and I guess I'd like to use the `SpanCategorizer` over `NER` for my use case, and would also like to ask if this particular use case with assembled pipelines is covered by the `SpanCategorizer`?

**Comment by polm** [comment, marked as answer]
This use case is one of the reasons the `SpanCategorizer` exists, so I'd definitely recommend using it. The easiest alternative, which is not great, would be having multiple pipelines. 

Note that the overwriting related options available to rule-based methods ([docs](https://spacy.io/usage/rule-based-matching#:~:text=to%20overwrite%20overlapping%20entities)) are related to this, but won't solve your problem.

Note that one reason this feature isn't already in spaCy is that practically speaking NER models work better with more information, so they can learn not only about features for a single class, but about interactions between classes. If possible I would recommend combining training data for your different models and seeing if you can learn something useful with them together.

**Reply by narayanacharya6** [reply]
Thanks for your response! We currently do achieve what we want with multiple models each hosting a single pipeline, but it does become painful to manage quite a few. I was hoping with the assembled pipeline we have one single model to manage. 

I have not tried training a single NER with all the training data we have but that I guess that will still not work since the unified NER will still make a single prediction per span(?). The different models we do have cater to a wide variety of domains and my intuition is that it will be difficult for the models to generalize across these different domains and we would not get the same performance in terms of aggregate metrics that we currently get. Having multiple labels per span is important to us because we have some subsequent pipelines in our setup that make use of the labels predicted for each NER to further enrich the doc with some useful information.

If I did wish to change this behavior by say sub-classing the `EntityRuler` class or another approach, can you help with a brief explanation of how I could achieve this? Is it simply a matter of overriding the `set_ents` or `set_annotations` method?

Also, when is v3.1, the possible release for `SpanCategorizer`, expected to be released?

**Reply by polm** [reply]
Ah, yeah, if your pipeline relies on having overlapping annotations then training on all the data together won't help you much. 

We don't have a specific workaround for this at the moment since we're working on the SpanCategorizer instead. Overriding `set_annotations` sounds like it might be a pretty straightforward way to do it, but that's originally defined in the Parser in Cython, so the annotations are probably not in the format you expect. 

We do not have a date for v3.1 but it's not that far out.

**Reply by narayanacharya6** [reply]
I don't have much experience with Cython so I guess I'll just wait for v3.1 then :) Thanks again for your time!

**Comment by narayanacharya6** [comment]
I had a possible workaround that I wanted to bounce off someone before proceeding. Will adding a lightweight component between adjacent NER pipes in the assembled pipeline that stores the entities recognized in a different extension (say `doc._.all_ents`) and clear the original `doc.ents` work? Will no `ents` set on the `Doc` object be enough for the next NER pipe to assign its recognized entities to the `doc.ents`? Anything I might be missing in this (obviously ugly) hack?

**Reply by narayanacharya6** [reply]
Ok, so this hack seems to work. I am not sure if this has some other dependencies that might be impacted because of setting the `doc.ents` to an empty tuple. Would appreciate some feedback/insight into this :)

```
from spacy.language import Language
from spacy.tokens import Doc


@Language.component("dafuq.ner_hack.v1")
def ner_hack_component(doc: Doc):
    if not Doc.has_extension('all_ents'):
        Doc.set_extension('all_ents', default=[])

    if doc.ents:
        doc._.all_ents.extend(doc.ents)
        doc.ents = tuple()

    return doc
```

**Reply by polm** [reply]
Nice idea! I think that should work.

One note about the `set_extension` call - if you call that outside the component that should be simpler than calling it inside the component. Also I think calling it multiple times is harmless.

**Reply by polm** [reply]
Ah, actually... Does that not cause any issues with the Token attributes like `ent_type_`? If you ran the code and it works then it should be fine, but if you have any code that uses token attributes instead of `doc.ents` it could cause inconsistencies.

**Reply by narayanacharya6** [reply]
Yes, since we do not use token attributes for `ent*` (at least for now), it does not impact us. But, if anyone else uses this, they need to be aware of the issue you've pointed out. The token attributes will continue to reflect the behavior that I've reported in the original question.

Also, I agree with your point for `set_extension`. This was me just doing a quick POC to confirm if what I think works or not :)
